Auth0/CredentialsManager.swift

```diff
@@ -0,0 +1,89 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import Foundation
+import SimpleKeychain
+
+/// Credentials management utility
+public struct CredentialsManager {
+
+    private let storage = A0SimpleKeychain()
+    private let storeKey = "credentials"
+    private let authentication: Authentication
+
+    /// Creates a new CredentialsManager instance
+    ///
+    /// - Parameters:
+    ///   - authentication: Auth0 authentication instance
+    public init(authentication: Authentication) {
+        self.authentication = authentication
+    }
+
+    /// Store credentials instance in keychain
+    ///
+    /// - Parameter credentials: credentials instance to store
+    /// - Returns: Bool outcome of success
+    public func store(_ credentials: Credentials) -> Bool {
+        return self.storage.setData(NSKeyedArchiver.archivedData(withRootObject: credentials), forKey: storeKey)
+    }
+
+    /// Retrieve stored credentials instance from keychain
+    ///
+    /// - Returns: Optional Credentials instance
+    public func retrieve() -> Credentials? {
+        guard let data = self.storage.data(forKey:self.storeKey) else { return nil }
+        return NSKeyedUnarchiver.unarchiveObject(with: data) as? Credentials
+    }
+
+    /// Retrieve credentials from keychain and yield new credentials if accessToken has expired
+    /// Otherwise the retrieved credentails will be returned as still valid
+    ///
+    /// More Information: https://auth0.com/docs/tokens/refresh-token
+    ///
+    /// ```
+    /// credentialsManager.retrieveAndRenewExpired {
+    ///    guard $0 == nil else { return }
+    ///    print($1)
+    /// }
+    /// ```
+    ///
+    /// - Parameters:
+    ///   - scope: scopes to request for the new tokens. By default is nil which will ask for the same ones requested during original Auth
+    ///   - callback: callback with the user's credentials or the cause of the error.
+    /// - Important: This method only works for a refresh token obtained after auth with OAuth 2.0 API Authorization.
+    public func retrieveCredentialsAutoRenew(withScope scope: String? = nil, callback: @escaping (CredentialsManagerError?, Credentials?) -> Void) {
```

Comment:
Just call it either `credentials(withScope:, callback:)` or `withCredentials(scope:, callback)` the idea is to use this one to try to get credentials to be used. If it fails the user should login otherwise it will always yield a valid token